### PR TITLE
Add a TF in-graph tokenizer for BERT

### PR DESCRIPTION
**Pull request by Rocketknight1**
Hey all! I've done some testing and the in-graph BERT tokenizer is now yielding the same outputs as our tokenizers, even for multi-part texts where we need to concatenate and get `token_type_ids` right. There's still several things left to do before this is ready, but I figured now is the time to lay it out and get some feedback!

Left to do:

- [x] Add input normalization
- [x] Is texts_a / texts_b the right way to handle inputs? Should it just be a multidimensional tensor?
- [x] Should this be a complete class rather than reading attributes from an existing tokenizer?
- [x] Add imports and maybe some kind of AutoModel to make this findable by users
- [x] Add dependency for tensorflow_text and import checks
- [x] Do we need to change the name? Most TF users will still want the normal tokenizers
- [x] Add tests, particularly one with a full model and one with saving to savedmodel, as these are main use cases
- [x] Currently always pads to max length - this should be an option
- [x] Add documentation
- [ ] Consider adding docs on how to add other TF tokenizers, so we can see if users want to add them once we have a framework in place?

### CODE REVIEW

File: src/transformers/models/bert/tokenization_bert_tf.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        super().__init__()
+        self.tf_tokenizer = FastBertTokenizer(vocab_list, token_out_type=tf.int64)
+        self.vocab_list = vocab_list
+        self.do_lower_case = do_lower_case
+        self.cls_token_id = cls_token_id
+        self.sep_token_id = sep_token_id
+        self.pad_token_id = pad_token_id
+
+    @classmethod
+    def from_tokenizer(cls, tokenizer: PreTrainedTokenizer):
```
Comment (sgugger) [inline src/transformers/models/bert/tokenization_bert_tf.py:29]: If it works for both slow and fast tokenizers, the class should be "PreTrainedTokenizerBase"

Also please use forward references to avoid doing circular imports just for the sake of type-hinting.

File: src/transformers/models/bert/tokenization_bert_tf.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+from tensorflow_text import FastBertTokenizer, case_fold_utf8, combine_segments, pad_model_inputs
+
+from ... import PreTrainedTokenizer
+
+
+class TFBertTokenizer(tf.keras.layers.Layer):
+    # TODO Input normalization?
+    # TODO Is texts_a / texts_b the right way to handle inputs? Should it just be a multidimensional tensor?
+    # TODO Should this be a more complete class rather than reading data from an existing tokenizer?
+    # TODO Add imports and maybe some kind of AutoModel to make this findable by users
```
Comment (sgugger) [inline src/transformers/models/bert/tokenization_bert_tf.py:14]: Wait until we have several classes before doing that I'd say.

File: src/transformers/models/bert/tokenization_bert_tf.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+from tensorflow_text import FastBertTokenizer, case_fold_utf8, combine_segments, pad_model_inputs
+
+from ... import PreTrainedTokenizer
+
+
+class TFBertTokenizer(tf.keras.layers.Layer):
+    # TODO Input normalization?
+    # TODO Is texts_a / texts_b the right way to handle inputs? Should it just be a multidimensional tensor?
+    # TODO Should this be a more complete class rather than reading data from an existing tokenizer?
```
Comment (sgugger) [inline src/transformers/models/bert/tokenization_bert_tf.py:13]: Maybe implement a from_pretrained?

File: src/transformers/models/bert/tokenization_bert_tf.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        )
+
+    def unpaired_tokenize(self, texts):
+        if self.do_lower_case:
+            texts = case_fold_utf8(texts)
+        else:
+            texts = tf.constant(texts)
+        return self.tf_tokenizer.tokenize(texts)
+
+    def call(self, texts_a, texts_b=None):
```
Comment (sgugger) [inline src/transformers/models/bert/tokenization_bert_tf.py:48]: Here is the API should ultimately be the same as the regular tokenizer, with options for padding/truncation/max length I think.

File: src/transformers/models/bert/tokenization_bert_tf.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+import tensorflow as tf
+
+from tensorflow_text import FastBertTokenizer, case_fold_utf8, combine_segments, pad_model_inputs
+
+from ... import PreTrainedTokenizer
+
+
+class TFBertTokenizer(tf.keras.layers.Layer):
+    # TODO Input normalization?
```
Comment (sgugger) [inline src/transformers/models/bert/tokenization_bert_tf.py:11]: Probably necessary to have the same results as with a traditional pipeline.

**Review by sgugger — COMMENTED** [review]
Great that you're working on this!

File: src/transformers/models/bert/tokenization_bert_tf.py (hunk 1 above)
Comment (Rocketknight1) [reply src/transformers/models/bert/tokenization_bert_tf.py:29]: Fixed!

File: src/transformers/models/bert/tokenization_bert_tf.py (hunk 3 above)
Comment (Rocketknight1) [reply src/transformers/models/bert/tokenization_bert_tf.py:14]: Done!

File: src/transformers/models/bert/tokenization_bert_tf.py (hunk 4 above)
Comment (Rocketknight1) [reply src/transformers/models/bert/tokenization_bert_tf.py]: Done!

**Review by gante — COMMENTED** [review]
Looks cool! (bear in mind -- I know very little about the insides of our tokenizers :D )

**Comment by Rocketknight1** [comment]
@sgugger @gante I'm happy with this PR now, so I'm ready for a final review! Also, the tests are slightly slow (~1min for the whole test suite on a single core). Should I mark some of them as `@slow`, or will they only be run nightly and when a PR affects the BERT directory anyway?

**Review by gante — APPROVED** [review]
LGTM 👍 

(I'd add the `@slow`, if the tests are indeed slow)

File: src/transformers/models/bert/tokenization_bert_tf.py [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+    """
+
+    def __init__(
+        self,
+        vocab_list: List,
+        do_lower_case: bool,
+        cls_token_id: int = None,
+        sep_token_id: int = None,
+        pad_token_id: int = None,
+        padding: str = "longest",
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply src/transformers/models/bert/tokenization_bert_tf.py:57]: Fixed!

File: src/transformers/models/bert/tokenization_bert_tf.py [hunk 7]
```diff
@@ -0,0 +1,220 @@
+import os
+from typing import List, Union
+
+import tensorflow as tf
+
+from tensorflow_text import FastBertTokenizer, ShrinkLongestTrimmer, case_fold_utf8, combine_segments, pad_model_inputs
+
+from ...utils import requires_backends
```
Comment (sgugger) [inline src/transformers/models/bert/tokenization_bert_tf.py:8]: You don't need this anymore since you went the dummy route. The dummy objects will be imported if tf text is not installed and raise the error for you.

File: src/transformers/models/bert/tokenization_bert_tf.py [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+        padding: str = "longest",
+        truncation: bool = True,
+        max_length: int = 512,
+        pad_to_multiple_of: int = None,
+        return_token_type_ids: bool = True,
+        return_attention_mask: bool = True,
+    ):
+        super().__init__()
+
+        requires_backends(self, ["tensorflow_text"])
```
Comment (sgugger) [inline src/transformers/models/bert/tokenization_bert_tf.py:69]:
```suggestion
```

File: src/transformers/models/bert/tokenization_bert_tf.py [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+
+        Examples:
+
+        ```python
+        from transformers import TFBertTokenizer
+
+        tf_tokenizer = TFBertTokenizer.from_pretrained("bert-base-uncased")
+        ```
+        """
+        tokenizer = BertTokenizer.from_pretrained(pretrained_model_name_or_path, *init_inputs, **kwargs)
```
Comment (sgugger) [inline src/transformers/models/bert/tokenization_bert_tf.py:133]: Probably best to try/except a `BertTokenizer` and then a `BertTokenizerFast` since there may be repos with only the fast tokenizer file available.

And the `BertTokenizerFast` import should be in the except to avoid the dependency on tokenizers.

**Review by sgugger — APPROVED** [review]
Nice work on this!

File: src/transformers/models/bert/tokenization_bert_tf.py (hunk 7 above)
Comment (Rocketknight1) [reply src/transformers/models/bert/tokenization_bert_tf.py:8]: I don't think I made the dummy objects in the end! I'll leave it in for now, but might switch to dummy objects when we add a few more tokenizers

File: src/transformers/models/bert/tokenization_bert_tf.py (hunk 9 above)
Comment (Rocketknight1) [reply src/transformers/models/bert/tokenization_bert_tf.py]: Done!

File: src/transformers/models/bert/tokenization_bert_tf.py (hunk 8 above)
Comment (Rocketknight1) [reply src/transformers/models/bert/tokenization_bert_tf.py:69]: Also skipping this because I don't think the dummy objects exist (but I'll take this advice later when I add them!)

File: src/transformers/models/bert/tokenization_bert_tf.py (hunk 7 above)
Comment (sgugger) [reply src/transformers/models/bert/tokenization_bert_tf.py:8]: This is not resolved yet, please do address :-)

Comment (Rocketknight1) [reply src/transformers/models/bert/tokenization_bert_tf.py:8]: Ah, sorry!

Comment (Rocketknight1) [reply src/transformers/models/bert/tokenization_bert_tf.py]: @sgugger I dug around in this one for a little bit - I figured out the reason I had trouble with the original approach with dummies is a missing `_` in one of the regexes in `check_dummies.py`, which caused it to miss `is_tensorflow_text_available`. Once that was fixed I was able to create `dummy_tensorflow_text_objects.py` without problems and remove `requires_backends` in the class `__init__()`, so hopefully everything is okay now! If you could double-check it to make sure I didn't miss anything, because I'm still a little uncertain about adding a new set of dummy objects like this, that'd be great!

Comment (sgugger) [reply src/transformers/models/bert/tokenization_bert_tf.py]: That works for me!

**Review by sgugger — APPROVED** [review]

File: utils/check_dummies.py [hunk 10]
```diff
@@ -26,7 +26,7 @@
 _re_backend = re.compile(r"is\_([a-z_]*)_available()")
 # Matches from xxx import bla
 _re_single_line_import = re.compile(r"\s+from\s+\S*\s+import\s+([^\(\s].*)\n")
-_re_test_backend = re.compile(r"^\s+if\s+not\s+is\_[a-z]*\_available\(\)")
+_re_test_backend = re.compile(r"^\s+if\s+not\s+is\_[a-z_]*\_available\(\)")
```
Comment (sgugger) [inline utils/check_dummies.py:29]: Nice catch!
